File: src/context.py
import abc
import asyncio
import logging
import time

logger = logging.getLogger(__name__)


class AsyncEventHandlerAdapterInterface(abc.ABC):
    @abc.abstractmethod
    async def start(self) -> None:
        raise NotImplementedError

    @abc.abstractmethod
    async def stop(self) -> None:
        raise NotImplementedError

    async def __aenter__(self, **kwargs):
        await self.start()
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        if exc_val:
            logger.warning(f'exc_type: {exc_type}')
            logger.warning(f'exc_value: {exc_val}')
            logger.warning(f'exc_traceback: {exc_tb}')
        await self.stop()


class AsyncEventHandlerThreadedAdapter(AsyncEventHandlerAdapterInterface):
    def __init__(self):
        self.thread = None
        self.thread_name = type(self)
        self.thread_running = False

    @abc.abstractmethod
    async def handle_event(self) -> None:
        raise NotImplementedError

    async def delay(self, interval: int) -> None:
        t0 = time.time()
        while self.thread_running and (t0 + interval > time.time()):
            await asyncio.sleep(1)
        return self.thread_running

    async def start(self):
        if not self.thread:
            self.thread = asyncio.create_task(self.handle_event())
            self.thread_running = True

    async def stop(self):
        if self.thread:
            # await self.before_stop()
            self.thread_running = False
            self.thread.cancel()
            try:
                await self.thread
            except asyncio.CancelledError:
                pass
            self.thread = None

File: src/context.js
const { EventEmitter } = require('events');

class AsyncEventHandlerAdapterInterface {
  async start() {
    throw new Error('Method not implemented.');
  }

  async stop() {
    throw new Error('Method not implemented.');
  }

  async __aenter__(kwargs) {
    await this.start();
    return this;
  }

  async __aexit__(exc_type, exc_val, exc_tb) {
    if (exc_val) {
      console.warning(`exc_type: ${exc_type}`);
      console.warning(`exc_value: ${exc_val}`);
      console.warning(`exc_traceback: ${exc_tb}`);
    }
    await this.stop();
  }
}

class AsyncEventHandlerThreadedAdapter extends AsyncEventHandlerAdapterInterface {
  constructor() {
    super();
    this.thread = null;
    this.thread_name = this.constructor.name;
    this.thread_running = false;
  }

  async handle_event() {
    throw new Error('Method not implemented.');
  }

  async delay(interval) {
    const t0 = Date.now();
    while (this.thread_running && t0 + interval > Date.now()) {
      await new Promise((resolve) => setTimeout(resolve, 1000));
    }
    return this.thread_running;
  }

  async start() {
    if (!this.thread) {
      this.thread = (async () => {
        try {
          await this.handle_event();
        } catch (error) {
          console.error(error);
        }
      })();
      this.thread_running = true;
    }
  }

  async stop() {
    if (this.thread) {
      this.thread_running = false;
      this.thread = null;
    }
  }
}
